src/main/java/htsjdk/io/AsyncWriterPool.java

```diff
@@ -0,0 +1,270 @@
+            if (this.currentTask != null && this.currentTask.isDone()) {
+                this.checkAndRethrow();
+            }
+        }
+
+        /**
+         * Blocking check of the previously completed task. This will wait for the future to complete if it is not null.
+         * If the task generated any exception, these will be re-thrown.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void blockingCheckAndRethrow() {
+            if (this.currentTask != null) {
+                this.checkAndRethrow();
+            }
+        }
+
+        /**
+         * Collect any exceptions thrown by the task, close writer, set the current task to null, and rethrow the
+         * caught exceptions.
+         * <p>
+         * This method should not be called directly and should only be used through the {@code blocking} and
+         * {@code nonBlocking} variants that go by the same name.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void checkAndRethrow() {
+            try {
+                this.currentTask.get();
+            } catch (CancellationException | InterruptedException | ExecutionException e) {
+                this.isClosed = true;
+                throw new RuntimeException("Exception while writing records asynchronously", e);
+            } finally {
+                // Set to null to avoid checking the same task multiple times
+                this.currentTask = null;
+            }
+        }
+
+        /**
+         * Add an item to the {@link PooledWriter}'s queue for writing. If the number of items in the queue exceeds the
+         * set {@code writeThreshold} and the previous task has been checked and is now null, the queue will be sent to
+         * the {@link AsyncWriterPool}'s executor for draining.
+         *
+         * @param item an item to enqueue for writing
+         * @throws RuntimeIOException if queue is already closed
+         * @throws RuntimeException   if the previous attempt to write failed, or if the attempt to add to the queue has
+         *                            been interrupted
+         */
+        @Override
+        public void write(final A item) {
+            if (this.isClosed) throw new RuntimeIOException("Attempt to add record to closed writer.");
+
+            this.nonBlockingCheckAndRethrow();
+
+            // Put new item in queue
+            try {
+                while (!this.isClosed && !this.queue.offer(item, 5, TimeUnit.SECONDS)) {
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Exception while placing item in queue", e);
+            }
+
+            // If queue size is large enough, and the there is no last task
+            if (queue.size() >= this.writeThreshold && this.currentTask == null) {
```

Comment:
```suggestion
            if (this.currentTask == null && queue.size() >= this.writeThreshold) {
```
Just inverting the order of checks since a null-check is fast and I think `size()` on blocking queues is a little more expensive.